src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperContext.cs

```diff
@@ -18,15 +18,19 @@ public class TagHelperContext
         /// Instantiates a new <see cref="TagHelperContext"/>.
         /// </summary>
         /// <param name="allAttributes">Every attribute associated with the current HTML element.</param>
+        /// <param name="items">TODO</param>
         /// <param name="uniqueId">The unique identifier for the source element this <see cref="TagHelperContext" /> 
         /// applies to.</param>
         /// <param name="getChildContentAsync">A delegate used to execute and retrieve the rendered child content 
         /// asynchronously.</param>
-        public TagHelperContext([NotNull] IDictionary<string, object> allAttributes,
-                                [NotNull] string uniqueId,
-                                [NotNull] Func<Task<string>> getChildContentAsync)
+        public TagHelperContext(
+            [NotNull] IDictionary<string, object> allAttributes,
+            [NotNull] IDictionary<string, object> items,
+            [NotNull] string uniqueId,
+            [NotNull] Func<Task<string>> getChildContentAsync)
         {
```

Comment:
shouldn't have callers deal with the intricacies of `CopyOnWriteDictionary`.  instead, do something similar to `ViewDataDictionary` and provide two constructors:
- a constructor for a new top-level context.  this one initializes `Items` with a regular old `Dictionary<string, object>(StringComparer.OrdinalIgnoreCase)`.
- a constructor that takes either the "parent" `TagHelperContext` (or "outer" `IDictionary<string, object>`).  this one initializes `Items` with `CopyOnWriteDictionary<string, object>(parent.Items, StringComparer.OrdinalIgnoreCase)`.